Add Cloudant terraform website docs (#3)
cloudant/terraform-provider-ibm: website/docs/r/cloudant.html.markdown

---
layout: "ibm"
page_title: "IBM : ibm_cloudant"
description: |-
  Manages Cloudant instance.

> subcategory: "Cloudant"

Reviewer: ```suggestion
subcategory: "Cloud Databases"
```

Author: Changed in ee4f0a13

  location = "us-south"
  plan     = "standard"

  legacy_credentials  = true
  include_data_events = false
  capacity            = 1
  enable_cors         = true

  cors_config {
    allow_credentials = false
    origins           = ["https://example.com"]
  }

  timeouts {
    create = "15m"
    update = "15m"
    delete = "15m"
  }
}
```

## Timeouts

ibm_cloudant provides the following [Timeouts](https://www.terraform.io/docs/language/resources/syntax.html#operation-timeouts) 
configuration options:

* `create` - (Default 10 minutes) The creation of the IBM Cloudant instance is considered failed if no response received.
* `delete` - (Default 10 minutes) The update of the IBM Cloudant instance is considered failed if no response received.
* `update` - (Default 10 minutes) The deletion of the IBM Cloudant instance is considered failed if no response received.

## Argument Reference

The following arguments are supported:

* `name` - (Required, String) A name for the resource instance.
* `location` - (Required, Forces new resource, String) Target location or environment to create the resource instance.
* `plan` - (Required, String) The plan type of the service.

> * `capacity` - (Optional, Number) A number of blocks of throughput units. A block consists of 100 reads/sec, 50 writes/sec, and 5 global queries/sec of provisioned throughput capacity.

Reviewer: As above I think we should link out for the exact numbers.

  * Constraints: The default value is `1`. Setting capacity is not supported for lite plan.
* `cors_config` - (Optional, Block List) Configuration for CORS. (see [below for nested arguments](#nestedblock--cors_config)) 
  * Constraints: The minimum length is `1` item. Can conflict with `enable_cors` if it is set to `false`. In this case the CORS customization is not allowed.
  * `origins` - (Required, List of String) An array of strings that contain allowed origin domains. You have to specify the full URL including the protocol. It is recommended that only the HTTPS protocol is used. Subdomains count as separate domains, so you have to specify all subdomains used.
  * `allow_credentials` - (Optional, Boolean) Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true. 
    * Constraints: The default value is `true`.
* `enable_cors` - (Optional, Boolean) Boolean value to turn CORS on and off.

>   * Constraints: The default value is `true`.If it is set to `false`, then customizing `cors_config` is not allowed.

Reviewer: nit: space
```suggestion
  * Constraints: The default value is `true`. If it is set to `false`, then customizing `cors_config` is not allowed.
```

Author: Changed in ee4f0a13

* `environment_crn` - (Optional, Forces new resource, String) CRN of the IBM Cloudant Dedicated Hardware plan instance.

> * `include_data_events` - (Optional, Boolean) Include data event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default only emitted events are of "management" type.

Reviewer: update to match suggestions from data source

Author: Changed in ee4f0a13

* `resource_plan_id` - (String) The unique ID of the plan associated with the offering.
* `resource_status` - (String) The status of the resource.
* `restored_at` - (String) The date when the instance under reclamation was restored.
* `restored_by` - (String) The subject who restored the instance back from reclamation.
* `scheduled_reclaim_at` - (String) The date when the instance was scheduled for reclamation.
* `scheduled_reclaim_by` - (String) The subject who initiated the instance reclamation.
* `service` - (String) The service type of the instance.
* `state` - (String) The current state of the instance.
* `status` - (String) Status of resource instance.
* `sub_type` - (String) The sub-type of instance, e.g. cfaas .
* `target_crn` - (String) The full deployment CRN as defined in the global catalog.
* `throughput` - (Map of Number) Schema for detailed information about throughput capacity with breakdown by specific throughput requests classes.
* `type` - (String) The type of the instance, e.g. service_instance.
* `update_at` - (String) The date when the instance was last updated.
* `update_by` - (String) The subject who updated the instance.

## Import

> You can import the `ibm_cloudant` resource by using `crn`. An array of event types that are being sent to IBM Cloud Activity Tracker for the IBM Cloudant instance. Allowable values: management, data. "management" is a required element of this array.

Reviewer: I think there is some copy/paste content left here?
```suggestion
You can import the `ibm_cloudant` resource by using `crn`.
```

Author: Changed in ee4f0a13

### Syntax

```hcl-terraform
$ terraform import ibm_cloudant.mycloudant <crn>
```

### Example
```hcl

> $ terraform import ibm_cloudant.mycloudant "crn:v1:bluemix:public:cloudantnosqldb:us-south:a/4ea2849a2d3401ed1e457268541966ea:79262bd4-4067-4783-b5ce-b1dba84ff8c4::"

Reviewer: Perhaps we should use something less real
```suggestion
$ terraform import ibm_cloudant.mycloudant "crn:v1:bluemix:public:cloudantnosqldb:us-south:a/abc123abc123abc123abc1:abc123ab-1234-1234-abc1-abc123abc123::"
```

Author: Changed in ee4f0a13

    allow_credentials = false
    origins           = ["https://example.com"]
  }

  timeouts {
    create = "15m"
    update = "15m"
    delete = "15m"
  }
}
```

## Timeouts

ibm_cloudant provides the following [Timeouts](https://www.terraform.io/docs/language/resources/syntax.html#operation-timeouts) 
configuration options:

* `create` - (Default 10 minutes) The creation of the IBM Cloudant instance is considered failed if no response received.
* `delete` - (Default 10 minutes) The update of the IBM Cloudant instance is considered failed if no response received.
* `update` - (Default 10 minutes) The deletion of the IBM Cloudant instance is considered failed if no response received.

## Argument Reference

The following arguments are supported:

* `name` - (Required, String) A name for the resource instance.
* `location` - (Required, Forces new resource, String) Target location or environment to create the resource instance.
* `plan` - (Required, String) The plan type of the service.
* `capacity` - (Optional, Number) A number of blocks of throughput units. A block consists of 100 reads/sec, 50 writes/sec, and 5 global queries/sec of provisioned throughput capacity.

>   * Constraints: The default value is `1`. Setting capacity is not supported for lite plan.

Reviewer 2: > Setting capacity is not supported for lite plan.

If I remember correctly it is, but cannot set to another value than 1.

Author: I think this is the same thing, just playing with words. Like: `half full or half empty glass`. In the end if there is `lite` plan, the capacity cannot be set/changed.

Reviewer: Should we format "lite" as `Lite` perhaps, as it is a specific plan name?

Reviewer: hmm I see actually the value we pass to the RC is `lite` so maybe we should leave the casing, but just put it in backticks.

Author: I changed it to `lite`: ee4f0a13

subcategory: "Cloudant"
---

# ibm_cloudant

Provides a resource for IBM Cloudant. This allows an IBM Cloudant service instance to be created, updated, or deleted.
For more information, about how to use this service, see the official [Cloudant documentation](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-getting-started-with-cloudant).

## Example Usage

```hcl
resource "ibm_cloudant" "cloudant" {
  name     = "cloudant-service-name"
  location = "us-south"
  plan     = "standard"

  legacy_credentials  = true
  include_data_events = false
  capacity            = 1
  enable_cors         = true

  cors_config {
    allow_credentials = false
    origins           = ["https://example.com"]
  }

  timeouts {
    create = "15m"
    update = "15m"
    delete = "15m"
  }
}
```

## Timeouts

ibm_cloudant provides the following [Timeouts](https://www.terraform.io/docs/language/resources/syntax.html#operation-timeouts) 
configuration options:

* `create` - The creation of the IBM Cloudant instance is considered failed if no response received for 10 minutes.
* `delete` - The update of the IBM Cloudant instance is considered failed if no response received for 10 minutes.

> * `update` - The deletion of the IBM Cloudant instance is considered failed if no response received for 10 minutes.

Reviewer 2: ```suggestion
* `create` - The creation of the IBM Cloudant instance is considered failed if no response received for 15 minutes.
* `delete` - The update of the IBM Cloudant instance is considered failed if no response received for 15 minutes.
* `update` - The deletion of the IBM Cloudant instance is considered failed if no response received for 15 minutes.
```

or update the timeouts block above to `10m`

Author: The default value is 10 minutes, but can be redefined with the timeout argument. I changed the text here: 574844eb

  * Constraints: The minimum length is `1` item.
* `enable_cors` - (Optional, Boolean) Boolean value to turn CORS on and off.
  * Constraints: The default value is `true`.
* `environment_crn` - (Optional, Forces new resource, String) CRN of the IBM Cloudant Dedicated Hardware plan instance.
* `hipaa` - (Optional, Forces new resource, Boolean) Instance is HIPAA ready in US locations.
  * Constraints: The default value is `false`.
* `include_data_events` - (Optional, Boolean) Include data event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default only emitted events are of "management" type.
  * Constraints: The default value is `false`.
* `kms_instance_crn` - (Optional, Forces new resource, String) CRN of the Key Protect instance housing the encryption key for BYOK.
* `kms_key_crn` - (Optional, Forces new resource, String) CRN of the encryption key that is stored in the Key Protect instance.
* `legacy_credentials` - (Optional, Forces new resource, Boolean) Use both legacy credentials and IAM for authentication.
  * Constraints: The default value is `false`.

> * `parameters` - (Optional, Forces new resource, Map) Arbitrary parameters to pass. Must be a JSON object.

Author: I am hesitant with this. I found it signed in [IBM Resource instance](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/resource_instance). But it is not signed in the [code](https://github.com/cloudant/terraform-provider-ibm/blob/4c138013b3025593c2def9da80ab426d653be1b8/ibm/resource_ibm_resource_instance.go#L89)
What to do with this question?

Reviewer: You mean the `ForceNew` part right?
There certainly appears to be a discrepency between the provider resource_instance docs and code.
@eiri WDYT? I'm inclined to leave this documented as forces new because I think the usage of it is likely to need it, but obviously the documentation means nothing if the code doesn't honour that.

Reviewer 3: I'd say let's follow `resource_instance` example and keep `Forces new resource`.

Factually this is incorrect, `parameters` work as "set once, ignore after" which is not something that can be expressed in terraform sdk, but we are delegating `parameters` management to `ibm_resource_instance`, so let's "delegate" deceision how to document it as well.

Author: Okay! Thanks! Then I leave it as-is now.
